src/rgw/rgw_rados.cc

```diff
@@ -8007,36 +8033,67 @@ int RGWRados::cls_bucket_list_ordered(RGWBucketInfo& bucket_info,
       ++count;
     }
 
-    // Refresh the candidates map
+    // refresh the candidates map
     candidates.erase(candidates.begin());
     ++vcurrents[pos];
     if (vcurrents[pos] != vends[pos]) {
       candidates[vcurrents[pos]->first] = pos;
-    }
-  }
+    } else if (list_results[pos].is_truncated) {
+      if (0 == additional_reads--) {
+	ldout(cct, 10) << "RGWRados::" << __func__ << ": returning " <<
+	  count << "entries early as we used all " <<
+	  allowed_additional_reads << " additional reads" << dendl;
+	break;
+      }
 
-  // Suggest updates if there is any
-  map<string, bufferlist>::iterator miter = updates.begin();
-  for (; miter != updates.end(); ++miter) {
-    if (miter->second.length()) {
+      ldout(cct, 10) << "RGWRados::" << __func__ <<
+	": issuing additional read for shard " << pos << "; " <<
+	additional_reads << " additional reads remaining" << dendl;
+
+      // we're at the end of this shard's list, but it's truncated, so
+      // read more in
+      map<int, string> shard_oid;
+      shard_oid[pos] = oids[pos];
+      // pass list_results in, as it will only overwrite entries
+      // corresponding to oids, and we only have one entry there
+      r = CLSRGWIssueBucketList(ioctx, name, prefix, num_entries_per_shard,
+				list_versions, shard_oid, list_results,
+				cct->_conf->rgw_bucket_index_max_aio)();
```

Comment:
if you're only fetching a single shard, can you use `cls_rgw_bucket_list_op()` instead of `CLSRGWIssueBucketList` the way `cls_bucket_list_unordered()` does?

Reply:
p.s. i'm skeptical about this part that fetches more from single shards, because we don't get any parallelism from these requests. we may be better off just returning a truncated listing to the client, and resuming the parallel listing on their next request

i feel like this part is orthogonal to the change that limits how many entries we request from each shard - can we investigate this in a separate pr?

Reply:
I would tend to agree--given the earlier proof that clients will send further requests

Reply:
If we are allowed to return truncated listings (and there aren't lots of misbehaving clients), it seems like that might be the way to go.   IE fetch 1/<num_shards> entries from each shard and return a mostly complete listing with no extra book keeping or re-fetching.

Reply:
So if we choose our "excess" per-shard # of entries well, per shard re-fetches should be very rare. When we do hit the end of one shard, if reading one shard a few times saves us from re-issuing reads to 2,000 shards (or 65,000 shards, which we allow), I think it would be worth it. Because if we return fewer than the requested # of entries, we're pretty much guaranteed that either the ultimate client or the internal list op will re-issue, hitting all shards.

So my proposal is to limit the # of per-shard re-reads to a small #, such as 2. I would also like to log these events at level 1, so we can get some data. Again, the re-reads and log entries should be rare. I know @cbodley is disinclined to to log extraneous stuff at level 1, but it can be labelled "INFO" so as not to raise concern from end-users.

Reply:
@cbodley wrote:
> we may be better off just returning a truncated listing to the client, and resuming the parallel listing on their next request

My thinking is starting to change towards this view, although I'd like to discuss more.

So the question is whether the parameter should be interpreted as # of desired entries or maximum # of entries the client can handle at a time.

The more typical use case is the desire to list the entire bucket, in which case no matter what's returned, it will be followed by additional requests until the end of the bucket list is reached. In that case the parameter is better interpreted the max that the client can handle at a time, in which case returning fewer is not a problem.

There may be use cases where the client wants *exactly* that number of items, in which case not returning that number guarantees a second call, but possibly no more. That seems wasteful in that it essentially doubles the work.

So now I'm thinking that adding padding of 25%-50% might be sufficient. I'll bet it returns what's asked for most of the time, but prevents us from adding a second call and therefore doubling the work in the second use case.

Reply:
I understood we'd confirmed that Amazon does not promise to return exactly any number of items, and even if it wanted to, there would be scenarios where it could not